chain/consensus/mir/state_manager.go

```diff
@@ -0,0 +1,215 @@
+}
+
+type StateManager struct {
+
+	// The current epoch number.
+	currentEpoch t.EpochNr
+
+	// For each epoch number, stores the corresponding membership.
+	memberships map[t.EpochNr]map[t.NodeID]t.NodeAddress
+
+	// Channel to send batches to Lotus.
+	NextBatch chan *Batch
+
+	// Channel to send a membership.
+	NewMembership chan map[t.NodeID]t.NodeAddress
+
+	MirManager *Manager
+
+	reconfigurationVotes map[t.EpochNr]map[string]int
+}
+
+func NewStateManager(initialMembership map[t.NodeID]t.NodeAddress, m *Manager) *StateManager {
+	// Initialize the membership for the first epochs.
+	// We use configOffset+2 memberships to account for:
+	// - The first epoch (epoch 0)
+	// - The configOffset epochs that already have a fixed membership (epochs 1 to configOffset)
+	// - The membership of the following epoch (configOffset+1) initialized with the same membership,
+	//   but potentially replaced during the first epoch (epoch 0) through special configuration requests.
+	memberships := make(map[t.EpochNr]map[t.NodeID]t.NodeAddress, ConfigOffset+2)
+	for e := 0; e < ConfigOffset+2; e++ {
+		memberships[t.EpochNr(e)] = initialMembership
+	}
+
+	sm := StateManager{
+		NextBatch:            make(chan *Batch),
+		NewMembership:        make(chan map[t.NodeID]t.NodeAddress, 1),
+		MirManager:           m,
+		memberships:          memberships,
+		currentEpoch:         0,
+		reconfigurationVotes: make(map[t.EpochNr]map[string]int),
+	}
+	return &sm
+}
+
+// RestoreState restores the application's state to the one represented by the passed argument.
+// The argument is a binary representation of the application state returned from Snapshot().
+func (sm *StateManager) RestoreState(snapshot []byte, config *commonpb.EpochConfig) error {
+	sm.currentEpoch = t.EpochNr(config.EpochNr)
+	sm.memberships = make(map[t.EpochNr]map[t.NodeID]t.NodeAddress, len(config.Memberships))
+
+	for e, membership := range config.Memberships {
+		sm.memberships[t.EpochNr(e)] = make(map[t.NodeID]t.NodeAddress)
+		for nID, nAddr := range membership.Membership {
+			var err error
+			sm.memberships[t.EpochNr(e)][t.NodeID(nID)], err = multiaddr.NewMultiaddr(nAddr)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	newMembership := maputil.Copy(sm.memberships[t.EpochNr(config.EpochNr+ConfigOffset)])
+	sm.memberships[t.EpochNr(config.EpochNr+ConfigOffset+1)] = newMembership
+
+	return nil
+}
+
+// ApplyTXs applies transactions received from the availability layer to the app state.
+func (sm *StateManager) ApplyTXs(txs []*requestpb.Request) error {
+	var msgs []Message
+
+	// For each request in the batch
+	for _, req := range txs {
+		switch req.Type {
+		case TransportType:
+			msgs = append(msgs, req.Data)
+		case ReconfigurationType:
+			err := sm.applyConfigMsg(req)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Send a batch to the Lotus node.
+	sm.NextBatch <- &Batch{
+		Messages:   msgs,
+		Validators: maputil.GetSortedKeys(sm.memberships[sm.currentEpoch]),
+	}
```

Comment:
Once snapshots are supported, this will need to happen synchronously. How hard would it be to forego the channel completely and do right here what the `mine` function does on reading from that channel?

In fact, maybe we could get rid of more of these channels and handle the functionality right in the state manager...